File: adapters/ruby/lib/lh_standard_adapter/dna_generator.rb
require "digest"
require "time"

module LhStandardAdapter
  TIAN_GAN = %w[Jia Yi Bing Ding Wu Ji Geng Xin Ren Gui].freeze
  DI_ZHI = %w[Zi Chou Yin Mao Chen Si Wu Wei Shen You Xu Hai].freeze
  SHI_CHEN = %w[ZiShi ChouShi YinShi MaoShi ChenShi SiShi WuShi WeiShi ShenShi YouShi XuShi HaiShi].freeze

  HEXAGRAMS = [
    {symbol:"䷀",en:"Qian",cn:"乾",domain:"governance"},
    {symbol:"䷁",en:"Kun",cn:"坤",domain:"archive"},
    {symbol:"䷂",en:"Zhun",cn:"屯",domain:"init"},
    {symbol:"䷃",en:"Meng",cn:"蒙",domain:"learn"},
    {symbol:"䷄",en:"Xu",cn:"需",domain:"async"},
    {symbol:"䷅",en:"Song",cn:"讼",domain:"legal"},
    {symbol:"䷜",en:"Kan",cn:"坎",domain:"engine"},
    {symbol:"䷝",en:"Li",cn:"离",domain:"audit"},
    {symbol:"䷲",en:"Zhen",cn:"震",domain:"security"},
    {symbol:"䷳",en:"Gen",cn:"艮",domain:"privacy"},
    {symbol:"䷸",en:"Xun",cn:"巽",domain:"deploy"},
    {symbol:"䷹",en:"Dui",cn:"兑",domain:"trust"},
    {symbol:"䷾",en:"JiJi",cn:"既济",domain:"complete"},
    {symbol:"䷿",en:"WeiJi",cn:"未济",domain:"progress"}
  ].freeze

  TASK_HEXAGRAM_MAP = {
    "default"=>"governance","code"=>"engine","deploy"=>"deploy","audit"=>"audit",
    "security"=>"security","archive"=>"archive","init"=>"init","learn"=>"learn",
    "legal"=>"legal","privacy"=>"privacy","trust"=>"trust","complete"=>"complete","progress"=>"progress"
  }.freeze

  class DNAGenerator
    CYCLE_YEAR = 1984
    CYCLE_MONTH = [2,4,6,8,10,0,2,4,6,8,10,0].freeze

    def initialize(uid: "9622", device: "HM-9622-001", locale: "Asia/Shanghai")
      @uid, @device, @locale = uid, device, locale
    end

    def generate(task_type: "default", action: "WRAP", version: nil)
      task_type = "default" if task_type.nil? || task_type.empty?
      action = "WRAP" if action.nil? || action.empty?
      version = "V1.0" if version.nil? || version.empty?

      now = Time.now.utc
      stem = compute_stem_branch(now)
      hex = select_hexagram(task_type)
      body = "ADAPTER-#{task_type.upcase}-#{action.upcase}-#{version}"
      raw = "#{stem[:year]}#{stem[:month]}#{stem[:day]}#{stem[:shichen]}#{hex[:symbol]}#{hex[:en]}#{body}#{@device}#{now.iso8601}"
      hash8 = Digest::SHA256.hexdigest(raw)[0,8]
      "#LongHun⚡️#{stem[:year]}·#{stem[:month]}·#{stem[:day]}·#{stem[:shichen]}·#{hex[:symbol]}#{hex[:en]}-#{body}-#{hash8}"
    end

    private

    def compute_stem_branch(dt)
      y, m, doy, h = dt.year, dt.month, dt.yday, dt.hour
      ys = (y - CYCLE_YEAR) % 10; yb = (y - CYCLE_YEAR) % 12
      mb = CYCLE_MONTH[(y - CYCLE_YEAR) % 10]
      ms = (mb + m - 1) % 10; mbr = (m + 1) % 12
      ds = (y - 1900 + (y - 1900) / 4 + doy) % 10
      db = (y - 1900 + (y - 1900) / 4 + doy) % 12
      si = h / 2

      { year: TIAN_GAN[ys] + DI_ZHI[yb],
        month: TIAN_GAN[ms] + DI_ZHI[mbr],
        day: TIAN_GAN[ds] + DI_ZHI[db],
        shichen: SHI_CHEN[si] }
    end

    def select_hexagram(task_type)
      domain = TASK_HEXAGRAM_MAP[task_type] || "governance"
      HEXAGRAMS.find { |h| h[:domain] == domain } || HEXAGRAMS.first
    end
  end
end

File: adapters/go/lhstandardadapter/dna_generator.go
package lhstandardadapter

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"
)

// Heavenly Stems (天干)
var tianGan = []string{"Jia", "Yi", "Bing", "Ding", "Wu", "Ji", "Geng", "Xin", "Ren", "Gui"}

// Earthly Branches (地支)
var diZhi = []string{"Zi", "Chou", "Yin", "Mao", "Chen", "Si",
	"Wu", "Wei", "Shen", "You", "Xu", "Hai"}

// Shi Chen (时辰) — 12 two-hour periods
var shiChen = []string{"ZiShi", "ChouShi", "YinShi", "MaoShi", "ChenShi", "SiShi",
	"WuShi", "WeiShi", "ShenShi", "YouShi", "XuShi", "HaiShi"}

// Hexagram represents an I Ching hexagram with metadata.
type Hexagram struct {
	Symbol string `json:"symbol"`
	EnName string `json:"en_name"`
	CnName string `json:"cn_name"`
	Domain string `json:"domain"`
}

var hexagrams = []Hexagram{
	{"䷀", "Qian", "乾", "governance"},
	{"䷁", "Kun", "坤", "archive"},
	{"䷂", "Zhun", "屯", "init"},
	{"䷃", "Meng", "蒙", "learn"},
	{"䷄", "Xu", "需", "async"},
	{"䷅", "Song", "讼", "legal"},
	{"䷜", "Kan", "坎", "engine"},
	{"䷝", "Li", "离", "audit"},
	{"䷲", "Zhen", "震", "security"},
	{"䷳", "Gen", "艮", "privacy"},
	{"䷸", "Xun", "巽", "deploy"},
	{"䷹", "Dui", "兑", "trust"},
	{"䷾", "JiJi", "既济", "complete"},
	{"䷿", "WeiJi", "未济", "progress"},
}

// Task-to-hexagram domain mapping
var taskHexagramMap = map[string]string{
	"default":  "governance",
	"code":     "engine",
	"deploy":   "deploy",
	"audit":    "audit",
	"security": "security",
	"archive":  "archive",
	"init":     "init",
	"learn":    "learn",
	"legal":    "legal",
	"privacy":  "privacy",
	"trust":    "trust",
	"complete": "complete",
	"progress": "progress",
}

// StemBranch holds the four pillars computed from a timestamp.
type StemBranch struct {
	Year    string `json:"year"`
	Month   string `json:"month"`
	Day     string `json:"day"`
	ShiChen string `json:"shichen"`
}

const (
	cycleYear  = 1984 // JiaZi year reference
	cstOffset  = 8    // UTC+8 for Asia/Shanghai
)

// cycleMonth maps year-stem-index to month stem offsets.
var cycleMonth = []int{2, 4, 6, 8, 10, 0, 2, 4, 6, 8}

// DNAGenerator generates v∞ format DNA traceability codes.
//
// Format: #LongHun⚡️{StemBranch}·{Hexagram}-{ModulePath}-{Hash8}
type DNAGenerator struct {
	UID    string
	Device string
	Locale string
}

// NewDNAGenerator creates a DNAGenerator with defaults.
func NewDNAGenerator(uid, device string) *DNAGenerator {
	return &DNAGenerator{UID: uid, Device: device, Locale: "Asia/Shanghai"}
}

// Generate produces a full DNA traceability string.
func (g *DNAGenerator) Generate(taskType, action, version string) string {
	if taskType == "" {
		taskType = "default"
	}
	if action == "" {
		action = "WRAP"
	}
	if version == "" {
		version = "V1.0"
	}

	now := time.Now().UTC().Add(cstOffset * time.Hour)
	sb := g.computeStemBranch(now)
	hg := g.selectHexagram(taskType)

	body := fmt.Sprintf("ADAPTER-%s-%s-%s", toUpper(taskType), toUpper(action), version)

	raw := fmt.Sprintf("%s%s%s%s%s%s%s%s%s",
		sb.Year, sb.Month, sb.Day, sb.ShiChen,
		hg.Symbol, hg.EnName,
		body,
		g.Device,
		now.Format(time.RFC3339Nano),
	)

	h := sha256.Sum256([]byte(raw))
	hash8 := hex.EncodeToString(h[:])[:8]

	return fmt.Sprintf("#LongHun⚡️%s·%s·%s·%s·%s%s-%s-%s",
		sb.Year, sb.Month, sb.Day, sb.ShiChen,
		hg.Symbol, hg.EnName,
		body, hash8)
}

// computeStemBranch computes Heavenly Stem + Earthly Branch for a time.
func (g *DNAGenerator) computeStemBranch(t time.Time) StemBranch {
	year := t.Year()

	yearStemIdx := mod(year-cycleYear, 10)
	yearBranchIdx := mod(year-cycleYear, 12)

	yearOffset := mod(year-cycleYear, 10)
	monthStemIdx := mod(cycleMonth[yearOffset]+int(t.Month())-1, 10)
	monthBranchIdx := mod(int(t.Month())+1, 12)

	dayOfYear := t.YearDay()
	dayBase := (year - 1900) + (year-1900)/4 + dayOfYear
	dayStemIdx := mod(dayBase, 10)
	dayBranchIdx := mod(dayBase, 12)

	shichenIdx := t.Hour() / 2

	return StemBranch{
		Year:    tianGan[yearStemIdx] + diZhi[yearBranchIdx],
		Month:   tianGan[monthStemIdx] + diZhi[monthBranchIdx],
		Day:     tianGan[dayStemIdx] + diZhi[dayBranchIdx],
		ShiChen: shiChen[shichenIdx],
	}
}

// selectHexagram selects an I Ching hexagram based on task type.
func (g *DNAGenerator) selectHexagram(taskType string) Hexagram {
	domain, ok := taskHexagramMap[taskType]
	if !ok {
		domain = "governance"
	}
	for _, h := range hexagrams {
		if h.Domain == domain {
			return h
		}
	}
	return hexagrams[0] // Default: Qian
}

// mod returns the non-negative remainder of a % b.
func mod(a, b int) int {
	m := a % b
	if m < 0 {
		m += b
	}
	return m
}

// toUpper returns the uppercase ASCII version of s.
func toUpper(s string) string {
	b := []byte(s)
	for i := range b {
		if b[i] >= 'a' && b[i] <= 'z' {
			b[i] -= 32
		}
	}
	return string(b)
}
